src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.util.concurrent.{ExecutorService, Executors}
+
+import scala.collection.mutable
+
+/**
+ * Pool of thread executors. There should be a 1-1 correspondence between C# threads
+ * and Java threads.
+ */
+object ThreadPool {
+
+  /**
+   * Map from threadId to corresponding executor.
+   */
+  private val executors: mutable.HashMap[Int, ExecutorService] =
+    new mutable.HashMap[Int, ExecutorService]()
+
+  /**
+   * Run some code on a particular thread.
+   *
+   * @param threadId Integer id of the thread.
+   * @param task Function to run on the thread.
+   */
+  def run(threadId: Int, task: () => Unit): Unit = {
+    val executor = getOrCreateExecutor(threadId)
+    val future = executor.submit(new Runnable {
+      override def run(): Unit = task()
```

Comment:
Question: if `task()` throws an exception, we get the same logging back to C# side, right?

Reply:
Yeah, all the logging and exception info is maintained on both sides.  `handleMethodCall()` still writes the exception info into the `DataOutputStream`.

```
Inner Exception 1:
JvmException: org.apache.spark.sql.AnalysisException: Path does not exist: file:/C:/Users/anfog/Source/sparknet/spark/artifacts/bin/Microsoft.Spark.Extensions.Delta/Debug/invalidpath;
	at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$org$apache$spark$sql$execution$datasources$DataSource$$checkAndGlobPathIfNecessary$1.apply(DataSource.scala:558)
	at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$org$apache$spark$sql$execution$datasources$DataSource$$checkAndGlobPathIfNecessary$1.apply(DataSource.scala:545)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.immutable.List.foreach(List.scala:392)
...
```

Reply:
thanks for verifying this!